File: pkg/ordered-map.go
package pkg

import (
	"fmt"
)

type color bool

const (
	RED   color = true
	BLACK color = false
)

type node[K comparable, V any] struct {
	key         K
	val         V
	left, right *node[K, V]
	color       color
	size        int
}

type RedBlackBST[K comparable, V any] struct {
	root *node[K, V]
}

func NewRedBlackBST[K comparable, V any]() *RedBlackBST[K, V] {
	return &RedBlackBST[K, V]{}
}

func (t *RedBlackBST[K, V]) isRed(x *node[K, V]) bool {
	if x == nil {
		return false
	}
	return x.color == RED
}

func (t *RedBlackBST[K, V]) size(x *node[K, V]) int {
	if x == nil {
		return 0
	}
	return x.size
}

func (t *RedBlackBST[K, V]) Size() int {
	return t.size(t.root)
}

func (t *RedBlackBST[K, V]) IsEmpty() bool {
	return t.root == nil
}

func (t *RedBlackBST[K, V]) Get(key K) (V, bool) {
	return t.get(t.root, key)
}

func (t *RedBlackBST[K, V]) get(x *node[K, V], key K) (V, bool) {
	for x != nil {
		switch {
		case key < x.key:
			x = x.left
		case key > x.key:
			x = x.right
		default:
			return x.val, true
		}
	}
	var zero V
	return zero, false
}

func (t *RedBlackBST[K, V]) Contains(key K) bool {
	_, found := t.Get(key)
	return found
}

func (t *RedBlackBST[K, V]) Put(key K, val V) {
	t.root = t.put(t.root, key, val)
	t.root.color = BLACK
}

func (t *RedBlackBST[K, V]) put(h *node[K, V], key K, val V) *node[K, V] {
	if h == nil {
		return &node[K, V]{key: key, val: val, color: RED, size: 1}
	}

	switch {
	case key < h.key:
		h.left = t.put(h.left, key, val)
	case key > h.key:
		h.right = t.put(h.right, key, val)
	default:
		h.val = val
	}

	if t.isRed(h.right) && !t.isRed(h.left) {
		h = t.rotateLeft(h)
	}
	if t.isRed(h.left) && t.isRed(h.left.left) {
		h = t.rotateRight(h)
	}
	if t.isRed(h.left) && t.isRed(h.right) {
		t.flipColors(h)
	}

	h.size = t.size(h.left) + t.size(h.right) + 1
	return h
}

func (t *RedBlackBST[K, V]) rotateRight(h *node[K, V]) *node[K, V] {
	x := h.left
	h.left = x.right
	x.right = h
	x.color = h.color
	h.color = RED
	x.size = h.size
	h.size = t.size(h.left) + t.size(h.right) + 1
	return x
}

func (t *RedBlackBST[K, V]) rotateLeft(h *node[K, V]) *node[K, V] {
	x := h.right
	h.right = x.left
	x.left = h
	x.color = h.color
	h.color = RED
	x.size = h.size
	h.size = t.size(h.left) + t.size(h.right) + 1
	return x
}

func (t *RedBlackBST[K, V]) flipColors(h *node[K, V]) {
	h.color = !h.color
	h.left.color = !h.left.color
	h.right.color = !h.right.color
}

func (t *RedBlackBST[K, V]) DeleteMin() {
	if t.IsEmpty() {
		panic("BST underflow")
	}

	if !t.isRed(t.root.left) && !t.isRed(t.root.right) {
		t.root.color = RED
	}

	t.root = t.deleteMin(t.root)
	if !t.IsEmpty() {
		t.root.color = BLACK
	}
}

func (t *RedBlackBST[K, V]) deleteMin(h *node[K, V]) *node[K, V] {
	if h.left == nil {
		return nil
	}

	if !t.isRed(h.left) && !t.isRed(h.left.left) {
		h = t.moveRedLeft(h)
	}

	h.left = t.deleteMin(h.left)
	return t.balance(h)
}

func (t *RedBlackBST[K, V]) DeleteMax() {
	if t.IsEmpty() {
		panic("BST underflow")
	}

	if !t.isRed(t.root.left) && !t.isRed(t.root.right) {
		t.root.color = RED
	}

	t.root = t.deleteMax(t.root)
	if !t.IsEmpty() {
		t.root.color = BLACK
	}
}

func (t *RedBlackBST[K, V]) deleteMax(h *node[K, V]) *node[K, V] {
	if t.isRed(h.left) {
		h = t.rotateRight(h)
	}

	if h.right == nil {
		return nil
	}

	if !t.isRed(h.right) && !t.isRed(h.right.left) {
		h = t.moveRedRight(h)
	}

	h.right = t.deleteMax(h.right)

	return t.balance(h)
}

func (t *RedBlackBST[K, V]) Delete(key K) {
	if !t.Contains(key) {
		return
	}

	if !t.isRed(t.root.left) && !t.isRed(t.root.right) {
		t.root.color = RED
	}

	t.root = t.delete(t.root, key)
	if !t.IsEmpty() {
		t.root.color = BLACK
	}
}

func (t *RedBlackBST[K, V]) delete(h *node[K, V], key K) *node[K, V] {
	if key < h.key {
		if !t.isRed(h.left) && !t.isRed(h.left.left) {
			h = t.moveRedLeft(h)
		}
		h.left = t.delete(h.left, key)
	} else {
		if t.isRed(h.left) {
			h = t.rotateRight(h)
		}
		if key == h.key && h.right == nil {
			return nil
		}
		if !t.isRed(h.right) && !t.isRed(h.right.left) {
			h = t.moveRedRight(h)
		}
		if key == h.key {
			x := t.min(h.right)
			h.key = x.key
			h.val = x.val
			h.right = t.deleteMin(h.right)
		} else {
			h.right = t.delete(h.right, key)
		}
	}
	return t.balance(h)
}

func (t *RedBlackBST[K, V]) moveRedLeft(h *node[K, V]) *node[K, V] {
	t.flipColors(h)
	if t.isRed(h.right.left) {
		h.right = t.rotateRight(h.right)
		h = t.rotateLeft(h)
		t.flipColors(h)
	}
	return h
}

func (t *RedBlackBST[K, V]) moveRedRight(h *node[K, V]) *node[K, V] {
	t.flipColors(h)
	if t.isRed(h.left.left) {
		h = t.rotateRight(h)
		t.flipColors(h)
	}
	return h
}

func (t *RedBlackBST[K, V]) balance(h *node[K, V]) *node[K, V] {
	if t.isRed(h.right) && !t.isRed(h.left) {
		h = t.rotateLeft(h)
	}
	if t.isRed(h.left) && t.isRed(h.left.left) {
		h = t.rotateRight(h)
	}
	if t.isRed(h.left) && t.isRed(h.right) {
		t.flipColors(h)
	}

	h.size = t.size(h.left) + t.size(h.right) + 1
	return h
}

func (t *RedBlackBST[K, V]) Height() int {
	return t.height(t.root)
}

func (t *RedBlackBST[K, V]) height(x *node[K, V]) int {
	if x == nil {
		return -1
	}
	return 1 + max(t.height(x.left), t.height(x.right))
}

func (t *RedBlackBST[K, V]) Min() (K, bool) {
	if t.IsEmpty() {
		var zero K
		return zero, false
	}
	return t.min(t.root).key, true
}

func (t *RedBlackBST[K, V]) min(x *node[K, V]) *node[K, V] {
	if x.left == nil {
		return x
	}
	return t.min(x.left)
}

func (t *RedBlackBST[K, V]) Max() (K, bool) {
	if t.IsEmpty() {
		var zero K
		return zero, false
	}
	return t.max(t.root).key, true
}

func (t *RedBlackBST[K, V]) max(x *node[K, V]) *node[K, V] {
	if x.right == nil {
		return x
	}
	return t.max(x.right)
}

func (t *RedBlackBST[K, V]) Floor(key K) (K, bool) {
	x := t.floor(t.root, key)
	if x == nil {
		var zero K
		return zero, false
	}
	return x.key, true
}

func (t *RedBlackBST[K, V]) floor(x *node[K, V], key K) *node[K, V] {
	if x == nil {
		return nil
	}
	if key == x.key {
		return x
	}
	if key < x.key {
		return t.floor(x.left, key)
	}
	y := t.floor(x.right, key)
	if y != nil {
		return y
	}
	return x
}

func (t *RedBlackBST[K, V]) Ceiling(key K) (K, bool) {
	x := t.ceiling(t.root, key)
	if x == nil {
		var zero K
		return zero, false
	}
	return x.key, true
}

func (t *RedBlackBST[K, V]) ceiling(x *node[K, V], key K) *node[K, V] {
	if x == nil {
		return nil
	}
	if key == x.key {
		return x
	}
	if key > x.key {
		return t.ceiling(x.right, key)
	}
	y := t.ceiling(x.left, key)
	if y != nil {
		return y
	}
	return x
}

func (t *RedBlackBST[K, V]) Select(rank int) (K, bool) {
	if rank < 0 || rank >= t.Size() {
		var zero K
		return zero, false
	}
	return t.select(t.root, rank).key, true
}

func (t *RedBlackBST[K, V]) select(x *node[K, V], rank int) *node[K, V] {
	if x == nil {
		return nil
	}
	leftSize := t.size(x.left)
	if leftSize > rank {
		return t.select(x.left, rank)
	} else if leftSize < rank {
		return t.select(x.right, rank-leftSize-1)
	} else {
		return x
	}
}

func (t *RedBlackBST[K, V]) Rank(key K) int {
	return t.rank(key, t.root)
}

func (t *RedBlackBST[K, V]) rank(key K, x *node[K, V]) int {
	if x == nil {
		return 0
	}
	if key < x.key {
		return t.rank(key, x.left)
	} else if key > x.key {
		return 1 + t.size(x.left) + t.rank(key, x.right)
	} else {
		return t.size(x.left)
	}
}

func (t *RedBlackBST[K, V]) Keys() []K {
	if t.IsEmpty() {
		return []K{}
	}
	min, _ := t.Min()
	max, _ := t.Max()
	return t.KeysInRange(min, max)
}

func (t *RedBlackBST[K, V]) KeysInRange(lo, hi K) []K {
	queue := make([]K, 0)
	t.keysInRange(t.root, &queue, lo, hi)
	return queue
}

func (t *RedBlackBST[K, V]) keysInRange(x *node[K, V], queue *[]K, lo, hi K) {
	if x == nil {
		return
	}
	cmplo := lo < x.key
	cmphi := hi > x.key
	if cmplo {
		t.keysInRange(x.left, queue, lo, hi)
	}
	if cmplo && cmphi {
		*queue = append(*queue, x.key)
	}
	if cmphi {
		t.keysInRange(x.right, queue, lo, hi)
	}
}

func (t *RedBlackBST[K, V]) SizeInRange(lo, hi K) int {
	if lo > hi {
		return 0
	}
	if t.Contains(hi) {
		return t.Rank(hi) - t.Rank(lo) + 1
	} else {
		return t.Rank(hi) - t.Rank(lo)
	}
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: rust/src/ordered_map.rs
use std::cmp::Ordering;

#[derive(Debug)]
enum Color {
    Red,
    Black,
}

#[derive(Debug)]
struct Node<K, V> {
    key: K,
    value: V,
    left: Option<Box<Node<K, V>>>,
    right: Option<Box<Node<K, V>>>,
    color: Color,
    size: usize,
}

impl<K, V> Node<K, V> {
    fn new(key: K, value: V, color: Color, size: usize) -> Self {
        Node {
            key,
            value,
            left: None,
            right: None,
            color,
            size,
        }
    }
}

pub struct OrderedMap<K, V> {
    root: Option<Box<Node<K, V>>>,
}

impl<K: Ord, V> OrderedMap<K, V> {
    pub fn new() -> Self {
        OrderedMap { root: None }
    }

    pub fn is_empty(&self) -> bool {
        self.size() == 0
    }

    pub fn size(&self) -> usize {
        self.size_node(&self.root)
    }

    fn size_node(&self, node: &Option<Box<Node<K, V>>>) -> usize {
        match node {
            Some(n) => n.size,
            None => 0,
        }
    }

    fn is_red(node: &Option<Box<Node<K, V>>>) -> bool {
        match node {
            Some(n) => matches!(n.color, Color::Red),
            None => false,
        }
    }

    pub fn get(&self, key: K) -> Option<&V> {
        self.get_node(&self.root, key)
    }

    fn get_node<'a>(&'a self, node: &'a Option<Box<Node<K, V>>>, key: K) -> Option<&'a V> {
        match node {
            Some(n) => match key.cmp(&n.key) {
                Ordering::Less => self.get_node(&n.left, key),
                Ordering::Greater => self.get_node(&n.right, key),
                Ordering::Equal => Some(&n.value),
            },
            None => None,
        }
    }

    pub fn put(&mut self, key: K, value: V) {
        self.root = Self::put_node(self.root.take(), key, value);
        if let Some(ref mut root) = self.root {
            root.color = Color::Black;
        }
    }

    fn put_node(node: Option<Box<Node<K, V>>>, key: K, value: V) -> Option<Box<Node<K, V>>> {
        let mut h = match node {
            Some(n) => n,
            None => return Some(Box::new(Node::new(key, value, Color::Red, 1))),
        };

        match key.cmp(&h.key) {
            Ordering::Less => h.left = Self::put_node(h.left.take(), key, value),
            Ordering::Greater => h.right = Self::put_node(h.right.take(), key, value),
            Ordering::Equal => h.value = value,
        }

        if Self::is_red(&h.right) && !Self::is_red(&h.left) {
            h = Self::rotate_left(h);
        }
        if Self::is_red(&h.left) && Self::is_red(&h.left.as_ref().unwrap().left) {
            h = Self::rotate_right(h);
        }
        if Self::is_red(&h.left) && Self::is_red(&h.right) {
            Self::flip_colors(&mut h);
        }

        h.size = Self::size_node(&h.left) + Self::size_node(&h.right) + 1;
        Some(h)
    }

    fn rotate_left(mut h: Box<Node<K, V>>) -> Box<Node<K, V>> {
        let mut x = h.right.take().unwrap();
        h.right = x.left.take();
        x.left = Some(h);
        x.color = x.left.as_ref().unwrap().color.clone();
        x.left.as_mut().unwrap().color = Color::Red;
        x.size = x.left.as_ref().unwrap().size;
        x.left.as_mut().unwrap().size = Self::size_node(&x.left.as_ref().unwrap().left)
            + Self::size_node(&x.left.as_ref().unwrap().right)
            + 1;
        x
    }

    fn rotate_right(mut h: Box<Node<K, V>>) -> Box<Node<K, V>> {
        let mut x = h.left.take().unwrap();
        h.left = x.right.take();
        x.right = Some(h);
        x.color = x.right.as_ref().unwrap().color.clone();
        x.right.as_mut().unwrap().color = Color::Red;
        x.size = x.right.as_ref().unwrap().size;
        x.right.as_mut().unwrap().size = Self::size_node(&x.right.as_ref().unwrap().left)
            + Self::size_node(&x.right.as_ref().unwrap().right)
            + 1;
        x
    }

    fn flip_colors(h: &mut Box<Node<K, V>>) {
        h.color = match h.color {
            Color::Red => Color::Black,
            Color::Black => Color::Red,
        };
        if let Some(ref mut left) = h.left {
            left.color = match left.color {
                Color::Red => Color::Black,
                Color::Black => Color::Red,
            };
        }
        if let Some(ref mut right) = h.right {
            right.color = match right.color {
                Color::Red => Color::Black,
                Color::Black => Color::Red,
            };
        }
    }

    pub fn delete_min(&mut self) {
        if self.is_empty() {
            panic!("OrderedMap underflow");
        }

        if !Self::is_red(&self.root.as_ref().unwrap().left)
            && !Self::is_red(&self.root.as_ref().unwrap().right)
        {
            self.root.as_mut().unwrap().color = Color::Red;
        }

        self.root = Self::delete_min_node(self.root.take());
        if !self.is_empty() {
            self.root.as_mut().unwrap().color = Color::Black;
        }
    }

    fn delete_min_node(mut h: Option<Box<Node<K, V>>>) -> Option<Box<Node<K, V>>> {
        if h.as_ref().unwrap().left.is_none() {
            return None;
        }

        if !Self::is_red(&h.as_ref().unwrap().left)
            && !Self::is_red(&h.as_ref().unwrap().left.as_ref().unwrap().left)
        {
            h = Some(Self::move_red_left(h.unwrap()));
        }

        h.as_mut().unwrap().left = Self::delete_min_node(h.as_mut().unwrap().left.take());
        Some(Self::balance(h.unwrap()))
    }

    pub fn delete_max(&mut self) {
        if self.is_empty() {
            panic!("OrderedMap underflow");
        }

        if !Self::is_red(&self.root.as_ref().unwrap().left)
            && !Self::is_red(&self.root.as_ref().unwrap().right)
        {
            self.root.as_mut().unwrap().color = Color::Red;
        }

        self.root = Self::delete_max_node(self.root.take());
        if !self.is_empty() {
            self.root.as_mut().unwrap().color = Color::Black;
        }
    }

    fn delete_max_node(mut h: Option<Box<Node<K, V>>>) -> Option<Box<Node<K, V>>> {
        if Self::is_red(&h.as_ref().unwrap().left) {
            h = Some(Self::rotate_right(h.unwrap()));
        }

        if h.as_ref().unwrap().right.is_none() {
            return None;
        }

        if !Self::is_red(&h.as_ref().unwrap().right)
            && !Self::is_red(&h.as_ref().unwrap().right.as_ref().unwrap().left)
        {
            h = Some(Self::move_red_right(h.unwrap()));
        }

        h.as_mut().unwrap().right = Self::delete_max_node(h.as_mut().unwrap().right.take());
        Some(Self::balance(h.unwrap()))
    }

    pub fn delete(&mut self, key: K) {
        if !self.contains(&key) {
            return;
        }

        if !Self::is_red(&self.root.as_ref().unwrap().left)
            && !Self::is_red(&self.root.as_ref().unwrap().right)
        {
            self.root.as_mut().unwrap().color = Color::Red;
        }

        self.root = Self::delete_node(self.root.take(), key);
        if !self.is_empty() {
            self.root.as_mut().unwrap().color = Color::Black;
        }
    }

    fn delete_node(mut h: Option<Box<Node<K, V>>>, key: K) -> Option<Box<Node<K, V>>> {
        if key < h.as_ref().unwrap().key {
            if !Self::is_red(&h.as_ref().unwrap().left)
                && !Self::is_red(&h.as_ref().unwrap().left.as_ref().unwrap().left)
            {
                h = Some(Self::move_red_left(h.unwrap()));
            }
            h.as_mut().unwrap().left = Self::delete_node(h.as_mut().unwrap().left.take(), key);
        } else {
            if Self::is_red(&h.as_ref().unwrap().left) {
                h = Some(Self::rotate_right(h.unwrap()));
            }
            if key == h.as_ref().unwrap().key && h.as_ref().unwrap().right.is_none() {
                return None;
            }
            if !Self::is_red(&h.as_ref().unwrap().right)
                && !Self::is_red(&h.as_ref().unwrap().right.as_ref().unwrap().left)
            {
                h = Some(Self::move_red_right(h.unwrap()));
            }
            if key == h.as_ref().unwrap().key {
                let x = Self::min_node(h.as_ref().unwrap().right.as_ref().unwrap());
                h.as_mut().unwrap().key = x.key.clone();
                h.as_mut().unwrap().value = x.value.clone();
                h.as_mut().unwrap().right =
                    Self::delete_min_node(h.as_mut().unwrap().right.take());
            } else {
                h.as_mut().unwrap().right = Self::delete_node(h.as_mut().unwrap().right.take(), key);
            }
        }
        Some(Self::balance(h.unwrap()))
    }

    fn move_red_left(mut h: Box<Node<K, V>>) -> Box<Node<K, V>> {
        Self::flip_colors(&mut h);
        if Self::is_red(&h.right.as_ref().unwrap().left) {
            h.right = Some(Self::rotate_right(h.right.take().unwrap()));
            h = Self::rotate_left(h);
            Self::flip_colors(&mut h);
        }
        h
    }

    fn move_red_right(mut h: Box<Node<K, V>>) -> Box<Node<K, V>> {
        Self::flip_colors(&mut h);
        if Self::is_red(&h.left.as_ref().unwrap().left) {
            h = Self::rotate_right(h);
            Self::flip_colors(&mut h);
        }
        h
    }

    fn balance(mut h: Box<Node<K, V>>) -> Box<Node<K, V>> {
        if Self::is_red(&h.right) && !Self::is_red(&h.left) {
            h = Self::rotate_left(h);
        }
        if Self::is_red(&h.left) && Self::is_red(&h.left.as_ref().unwrap().left) {
            h = Self::rotate_right(h);
        }
        if Self::is_red(&h.left) && Self::is_red(&h.right) {
            Self::flip_colors(&mut h);
        }

        h.size = Self::size_node(&h.left) + Self::size_node(&h.right) + 1;
        h
    }

    fn min_node(x: &Box<Node<K, V>>) -> &Box<Node<K, V>> {
        match &x.left {
            Some(left) => Self::min_node(left),
            None => x,
        }
    }

    fn max_node(x: &Box<Node<K, V>>) -> &Box<Node<K, V>> {
        match &x.right {
            Some(right) => Self::max_node(right),
            None => x,
        }
    }

    pub fn min(&self) -> Option<&K> {
        self.root.as_ref().map(|node| &Self::min_node(node).key)
    }

    pub fn max(&self) -> Option<&K> {
        self.root.as_ref().map(|node| &Self::max_node(node).key)
    }

    pub fn keys(&self) -> Vec<&K> {
        let mut vec = Vec::new();
        self.keys_in_order(&self.root, &mut vec);
        vec
    }

    fn keys_in_order<'a>(&'a self, node: &'a Option<Box<Node<K, V>>>, vec: &mut Vec<&'a K>) {
        if let Some(n) = node {
            self.keys_in_order(&n.left, vec);
            vec.push(&n.key);
            self.keys_in_order(&n.right, vec);
        }
    }
}
